File: modules/paralleization/tutorial.openai-agents.ts
/**
 * Run:
 * cd modules && node --experimental-strip-types ./paralleization/tutorial.openai-agents.ts
 *
 * 학습 목표:
 * 1) OpenAIProvider/Runner 기본 연결 구조 이해
 * 2) 역할이 다른 Agent를 분리해 구성하는 방법 학습
 * 3) Promise.all로 병렬 실행 후 결과를 구조화하는 방법 학습
 */
import {
  Agent,
  OpenAIProvider,
  Runner,
  setDefaultOpenAIKey,
  setOpenAIAPI,
} from "@openai/agents";

import { isDirectExecution } from "./tutorial.execution.ts";
import { type Logger, type TutorialStepResult } from "./tutorial.types.ts";

const STEP_ID: string = "step-3-openai-agents";
const LIBRARY: "openai-agents" = "openai-agents";
const OPENAI_MODEL: string = "gpt-4o-mini";
const SYNTHETIC_MODEL: string = "hf:moonshotai/Kimi-K2.5";
const SYNTHETIC_BASE_URL: string = "https://api.synthetic.new/openai/v1";

type ProviderConfig = {
  apiKey: string;
  modelName: string;
  baseUrl?: string;
};

/**
 * step 실행 결과를 동일 스키마로 맞춰 인덱스 튜토리얼에서 재사용합니다.
 */
function toStepResult(
  status: TutorialStepResult["status"],
  message: string,
  output?: string,
): TutorialStepResult {
  const result: TutorialStepResult = {
    stepId: STEP_ID,
    library: LIBRARY,
    status,
    message,
    output,
  };
  return result;
}

/**
 * SDK 반환 타입이 문자열/객체로 달라도
 * 튜토리얼 출력은 사람이 읽기 쉬운 문자열 형태로 통일합니다.
 */
function toText(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }

  return JSON.stringify(value, null, 2);
}

/**
 * OpenAI Agents SDK 결과에서 핵심 출력(finalOutput)을 안전하게 추출합니다.
 */
function extractOutput(agentResult: unknown): string {
  if (typeof agentResult !== "object" || agentResult === null) {
    return toText(agentResult);
  }

  if (!("finalOutput" in agentResult)) {
    return toText(agentResult);
  }

  const finalOutput: unknown = (agentResult as { finalOutput: unknown })
    .finalOutput;
  return toText(finalOutput);
}

/**
 * 키 입력은 최소화합니다.
 * - OPENAI_API_KEY가 있으면 OpenAI 기본 경로
 * - 없으면 SYNTHETIC_API_KEY + 고정 base URL 경로
 */
function resolveProviderConfig(): ProviderConfig | undefined {
  const openAiApiKey: string | undefined = process.env.OPENAI_API_KEY?.trim();
  if (openAiApiKey) {
    return {
      apiKey: openAiApiKey,
      modelName: OPENAI_MODEL,
    };
  }

  const syntheticApiKey: string | undefined = process.env.SYNTHETIC_API_KEY?.trim();
  if (syntheticApiKey) {
    return {
      apiKey: syntheticApiKey,
      modelName: SYNTHETIC_MODEL,
      baseUrl: SYNTHETIC_BASE_URL,
    };
  }

  return undefined;
}

/**
 * 모든 튜토리얼이 같은 STEP 로그 형태를 유지하면
 * 라이브러리별 학습 비교가 쉬워집니다.
 */
function logStep(log: Logger, stepNo: number, title: string): void {
  log(`[openai-agents] STEP ${stepNo}/3 ${title}`);
}

export async function runOpenAIAgentsStep(
  log: Logger = console.log,
): Promise<TutorialStepResult> {
  // Step 1) 실행 준비
  logStep(log, 1, "setup");
  const providerConfig: ProviderConfig | undefined = resolveProviderConfig();
  if (!providerConfig) {
    return toStepResult(
      "skipped",
      "OPENAI_API_KEY or SYNTHETIC_API_KEY is missing.",
    );
  }

  // Step 2) Provider/Runner/Agent 구성
  logStep(log, 2, "build two specialist agents");
  const provider: OpenAIProvider = new OpenAIProvider({
    apiKey: providerConfig.apiKey,
    baseURL: providerConfig.baseUrl,
    useResponses: false,
  });

  setDefaultOpenAIKey(providerConfig.apiKey);
  setOpenAIAPI("chat_completions");

  const runner: Runner = new Runner({
    modelProvider: provider,
    tracingDisabled: true,
  });

  const biologyAgent: Agent = new Agent({
    name: "BiologyExpert",
    model: providerConfig.modelName,
    instructions:
      "You are a biology expert. Summarize CRISPR technology in one sentence.",
  });
  const chemistryAgent: Agent = new Agent({
    name: "ChemistryExpert",
    model: providerConfig.modelName,
    instructions:
      "You are a chemistry expert. Summarize solid-state batteries in one sentence.",
  });

  // Step 3) 병렬 실행 + 출력 병합
  logStep(log, 3, "execute in parallel with Promise.all");
  const [biologyResult, chemistryResult]: [unknown, unknown] =
    await Promise.all([
      runner.run(biologyAgent, "Provide your summary.", { maxTurns: 3 }),
      runner.run(chemistryAgent, "Provide your summary.", { maxTurns: 3 }),
    ]);

  const output: string = JSON.stringify(
    {
      biology: extractOutput(biologyResult),
      chemistry: extractOutput(chemistryResult),
    },
    null,
    2,
  );

  return toStepResult(
    "completed",
    "OpenAI Agents tutorial completed.",
    output,
  );
}

const isNodeTestContext: boolean = process.env.NODE_TEST_CONTEXT !== undefined;
if (!isNodeTestContext && isDirectExecution(import.meta.url, process.argv)) {
  void runOpenAIAgentsStep().then((result: TutorialStepResult): void => {
    console.log(JSON.stringify(result, null, 2));
  }).catch((error: unknown): void => {
    const errorMessage: string =
      error instanceof Error ? error.message : String(error);
    console.error(errorMessage);
    process.exitCode = 1;
  });
}

File: modules/paralleization/tutorial_openai_agents.py
"""
tutorial_openai_agents.py

학습 목표:
1) openai-agents Python SDK의 Provider/Runner 구성 이해
2) 역할 분리된 Agent 두 개를 병렬 실행하는 방법 학습
3) 실행 결과(final_output)를 안전하게 텍스트로 정규화하는 방법 학습
"""

import asyncio
import json
import os
from typing import Any, Final

from tutorial_types import StepLogger, TutorialStepResult

STEP_ID: Final[str] = "step-3-openai-agents"
LIBRARY: Final[str] = "openai-agents"
OPENAI_MODEL: Final[str] = "gpt-4o-mini"
SYNTHETIC_MODEL: Final[str] = "hf:moonshotai/Kimi-K2.5"
SYNTHETIC_BASE_URL: Final[str] = "https://api.synthetic.new/openai/v1"


def _log_step(log: StepLogger, step_number: int, title: str) -> None:
    """공통 STEP 로그 포맷."""
    log(f"[openai-agents] STEP {step_number}/3 {title}")


def _to_text(value: Any) -> str:
    """튜토리얼 출력 형식을 문자열로 통일합니다."""
    if isinstance(value, str):
        return value

    return json.dumps(value, ensure_ascii=True, indent=2)


def _resolve_provider_config() -> tuple[str, str, str | None] | None:
    """키 입력은 최소화하고, 모델/URL은 학습용 상수로 고정합니다."""
    openai_api_key: str = os.getenv("OPENAI_API_KEY", "").strip()
    if openai_api_key:
        return openai_api_key, OPENAI_MODEL, None

    synthetic_api_key: str = os.getenv("SYNTHETIC_API_KEY", "").strip()
    if synthetic_api_key:
        return synthetic_api_key, SYNTHETIC_MODEL, SYNTHETIC_BASE_URL

    return None


async def run_openai_agents_tutorial(log: StepLogger = print) -> TutorialStepResult:
    """OpenAI Agents 단계별 튜토리얼 실행."""
    # Step 1) 실행 준비
    _log_step(log, 1, "setup")
    provider_config: tuple[str, str, str | None] | None = _resolve_provider_config()
    if provider_config is None:
        return TutorialStepResult(
            step_id=STEP_ID,
            library="openai-agents",
            status="skipped",
            message="OPENAI_API_KEY or SYNTHETIC_API_KEY is missing.",
        )

    try:
        from agents import (
            Agent,
            OpenAIProvider,
            RunConfig,
            Runner,
            set_default_openai_api,
            set_default_openai_key,
        )
    except Exception as exception:  # noqa: BLE001
        return TutorialStepResult(
            step_id=STEP_ID,
            library="openai-agents",
            status="skipped",
            message=(
                "openai-agents package is not available. "
                f"Install requirement first. detail={exception}"
            ),
        )

    api_key: str
    model_name: str
    base_url: str | None
    api_key, model_name, base_url = provider_config

    # Step 2) Provider/Runner/Agent 구성
    _log_step(log, 2, "build two specialist agents")
    set_default_openai_key(api_key)
    set_default_openai_api("chat_completions")

    model_provider: OpenAIProvider = OpenAIProvider(
        api_key=api_key,
        base_url=base_url,
        use_responses=False,
    )
    run_config: RunConfig = RunConfig(model_provider=model_provider)

    biology_agent: Agent = Agent(
        name="BiologyExpert",
        model=model_name,
        instructions="You are a biology expert. Summarize CRISPR in one sentence.",
    )
    chemistry_agent: Agent = Agent(
        name="ChemistryExpert",
        model=model_name,
        instructions=(
            "You are a chemistry expert. "
            "Summarize solid-state batteries in one sentence."
        ),
    )

    # Step 3) 병렬 실행 + 결과 병합
    _log_step(log, 3, "execute in parallel with asyncio.gather")
    try:
        biology_task: asyncio.Task[Any] = asyncio.create_task(
            Runner.run(
                biology_agent,
                "Provide your summary.",
                max_turns=3,
                run_config=run_config,
            )
        )
        chemistry_task: asyncio.Task[Any] = asyncio.create_task(
            Runner.run(
                chemistry_agent,
                "Provide your summary.",
                max_turns=3,
                run_config=run_config,
            )
        )
        biology_result: Any
        chemistry_result: Any
        biology_result, chemistry_result = await asyncio.gather(
            biology_task,
            chemistry_task,
        )

        biology_output: Any = biology_result.final_output
        chemistry_output: Any = chemistry_result.final_output
        output: str = json.dumps(
            {
                "biology": _to_text(biology_output),
                "chemistry": _to_text(chemistry_output),
            },
            ensure_ascii=True,
            indent=2,
        )
        return TutorialStepResult(
            step_id=STEP_ID,
            library="openai-agents",
            status="completed",
            message="OpenAI Agents tutorial completed.",
            output=output,
        )
    except Exception as exception:  # noqa: BLE001
        return TutorialStepResult(
            step_id=STEP_ID,
            library="openai-agents",
            status="failed",
            message=str(exception),
        )


async def _main() -> None:
    """직접 실행 시 단일 step 결과를 JSON으로 출력합니다."""
    result: TutorialStepResult = await run_openai_agents_tutorial()
    print(json.dumps(result.__dict__, indent=2, ensure_ascii=True))


if __name__ == "__main__":
    asyncio.run(_main())
